aws_lambda_powertools/utilities/idempotency/idempotency.py

```diff
@@ -0,0 +1,96 @@
+"""
+Primary interface for idempotent Lambda functions utility
+"""
+import logging
+from typing import Any, Callable, Dict
+
+from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+
+from ..typing import LambdaContext
+from .exceptions import AlreadyInProgressError, ItemNotFoundError
+from .persistence import STATUS_CONSTANTS, BasePersistenceLayer
+
+logger = logging.getLogger(__name__)
+
+
+def default_error_callback():
+    raise
+
+
+@lambda_handler_decorator
+def idempotent(
+    handler: Callable[[Any, LambdaContext], Any],
+    event: Dict[str, Any],
+    context: LambdaContext,
+    persistence: BasePersistenceLayer,
+) -> Any:
+    """
+    Middleware to handle idempotency
+
+    Parameters
+    ----------
+    handler: Callable
+        Lambda's handler
+    event: Dict
+        Lambda's Event
+    context: Dict
+        Lambda's Context
+    persistence: BasePersistenceLayer
+        Instance of BasePersistenceLayer to store data
+
+    Examples
+    --------
+    **Processes Lambda's event in an idempotent manner**
+        >>> from aws_lambda_powertools.utilities.idempotency import idempotent, DynamoDBPersistenceLayer
+        >>>
+        >>> persistence_store = DynamoDBPersistenceLayer(event_key="body", table_name="idempotency_store")
+        >>>
+        >>> @idempotent(persistence=persistence_store)
+        >>> def handler(event, context):
+        >>>     return {"StatusCode": 200}
+    """
+
+    persistence_instance = persistence
+    try:
+        event_record = persistence_instance.get_record(event)
+    except ItemNotFoundError:
+        persistence_instance.save_inprogress(event=event)
```

Comment:
@cakepietoast - A ItemAlreadyExistsError may be raised here, and it is not really a bad thing. 

For example 2 calls are made close to each other. The first call has not yet put in progress record and the second call gets ItemNotFoundError and tries to also put in a record.